image-filters/image-filters.markdown

```diff
@@ -395,26 +391,26 @@ public void keyPressed() {
 
 If you run the app again, every time you press a key it will output it to the
 console. Later, you’ll want to do different things depending on what key was
-pressed, and to do this you just switch on the key value (this exists in the
-`PApplet` superclass, and contains the last key pressed). 
+pressed, and to do this you just switch on the key value. (This exists in the
+`PApplet` superclass, and contains the last key pressed.) 
 
 
 ## Writing Tests 
 
 This app doesn’t do a lot yet, but we can already see number of places where
-things can go wrong, for example triggering the wrong action with key presses.
+things can go wrong; for example, triggering the wrong action with key presses.
 As we add complexity, we add more potential problems, such as updating the
-image state incorrectly, or miscalculations of the pixel colors after applying
-a filter. I also just (some think weirdly) enjoy writing unit tests. Whilst
+image state incorrectly, or miscalculating pixel colors after applying
+a filter. I also just enjoy (some think weirdly) writing unit tests. Whilst
 some people seem to think of testing as a thing that delays checking code in, I
-see tests as my #1 debugging tool, and an opportunity to deeply understand what
+see tests as my #1 debugging tool, and as an opportunity to deeply understand what
 is going on in my code.
 
-I adore Processing, but as covered above it’s designed as a tool for artists to
```

Comment:
Processing was described several times, so I tried to streamline the descriptions so as to reduce repetition.